Don't panic if the lhs of a div by zero is not statically known

oli-obk opened the pull request:
Fixes #73993 for real this time

r? @wesleywiser

oli-obk committed:
Don't panic if the lhs of a div by zero is not statically known

wesleywiser commented:
Thanks @oli-obk!

@bors r+ rollup

bors commented:
:pushpin: Commit f689839ab07b1575e355bf24dad2d35bbea99b68 has been approved by `wesleywiser`

<!-- @bors r=wesleywiser f689839ab07b1575e355bf24dad2d35bbea99b68 -->
<!-- homu: {"type":"Approved","sha":"f689839ab07b1575e355bf24dad2d35bbea99b68","approver":"wesleywiser"} -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 30, this pull request will be tested once the tree is reopened

oli-obk commented:
@bors r=wesleywiser

I rebased, there was a weird CI failure about the issue I added, not sure what's up with that.

bors commented:
:pushpin: Commit d512a6c5e0273c680ca32086eb8264ac58796da6 has been approved by `wesleywiser`

<!-- @bors r=wesleywiser d512a6c5e0273c680ca32086eb8264ac58796da6 -->
<!-- homu: {"type":"Approved","sha":"d512a6c5e0273c680ca32086eb8264ac58796da6","approver":"wesleywiser"} -->

Manishearth commented:
@bors r-

https://github.com/rust-lang/rust/pull/74306#issuecomment-657871025

eddyb commented:
@Manishearth this PR had already failed PR CI by the time the rollup was created, could we have bors gate on PR, and r- on failure?

Manishearth commented:
@eddyb cc @pietroalbini 
we sometimes do have spurious failures though

oli-obk commented:
Uhmm... are we running all tests twice on CI? https://dev.azure.com/rust-lang/rust/_build/results?buildId=34443&view=logs&j=698b0e4a-0480-533d-81fa-f38599aa328e&t=76a462a9-00ba-550c-bb03-47e063adb169&l=1000 already ran all tests and my test even passed, then later it failed when being run again with `--emit metadata`, which it shouldn't do, because I have `// build-pass`

oli-obk commented:
I can now at least reproduce with `--pass=check`, but *whyy*

wesleywiser commented:
Is `ConstProp` not actually running in check mode or something?

oli-obk commented:
The entire test should not be run in check mode. That's the whole point of `// build-pass` afaik. 

Though... you're right in asking, why did I even add `// build-pass`. I just added it because otherwise the test didn't pass. looking at the log, it seems like it's indeed not being run at all. Yea, so running with `--emit=metadata` causes const prop to be ignored entirely.

oli-obk commented:
Jop, if I make the function `const fn` (and thus causing its MIR to get written to metadata, the test passes and I see const prop in the log).

oli-obk commented:
I gotta run, pushing a hack commit that makes it work, but I gotta investigate.

RalfJung commented on src/librustc_mir/transform/const_prop.rs:
```diff
@@ -1004,11 +1004,27 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
                     let expected = ScalarMaybeUninit::from(Scalar::from_bool(*expected));
                     let value_const = self.ecx.read_scalar(value).unwrap();
                     if expected != value_const {
+                        enum DbgVal<T> {
```
This looks like something more generally useful that shouldn't be 5 layers deep inside a private function?

oli-obk replied:
it's very specifically tailored to this case. I'm sure we could come up with something more general and put it into `rustc_data_structures`

oli-obk commented:
Investigated. Running optimizations with `--emit=metadata` will slow down `cargo check`, it's unforntunate we have differences between `cargo check` and `cargo build`, but these are only related to the optimizer failing to optimize because your code has bugs (instead of it failing to optimize silently like every other optimizer I know - to be fair, I don't know many).

We have modified lots of tests so that check-only mode will not fail them, so let's get this PR moving, it fixes a nightly bug that will become a beta bug soon.

I squashed the hack commit into the main commit.

@bors r=wesleywiser

bors commented:
:pushpin: Commit 598d1cd4640387bfd257541d3a6d5a3e81906eb0 has been approved by `wesleywiser`

<!-- @bors r=wesleywiser 598d1cd4640387bfd257541d3a6d5a3e81906eb0 -->
<!-- homu: {"type":"Approved","sha":"598d1cd4640387bfd257541d3a6d5a3e81906eb0","approver":"wesleywiser"} -->

wesleywiser commented:
Looks like we just missed the beta train? #74323

oli-obk commented:
@bors r=wesleywiser

bors commented:
:pushpin: Commit 703f6803db4b2014486a1c7230bc965fd75a672e has been approved by `wesleywiser`

<!-- @bors r=wesleywiser 703f6803db4b2014486a1c7230bc965fd75a672e -->
<!-- homu: {"type":"Approved","sha":"703f6803db4b2014486a1c7230bc965fd75a672e","approver":"wesleywiser"} -->

spastorino commented:
Adding `T-compiler` label so we can go through this nomination during `T-compiler` meeting.

pnkfelix commented:
[discussed](https://zulip-archive.rust-lang.org/238009tcompilermeetings/31818weeklymeeting2020072354818.html#204801939) at weekly T-compiler triage meeting.

beta-accepted.
